design/view-change.md

```diff
@@ -0,0 +1,229 @@
+  with same _k_ in different views.
+  
+  After updating _P_ and _Q_ replica clears its log and multicasts message
+  _VIEW-CHANGE(i, v+1, h, C, P, Q)_, where _C_ is a set of _(ck, cd)_ tuples.
+  
+- Upon receiving _VIEW-CHANGE_ message replica checks whether all tuples
+  in _P_ and _Q_ are for view _v_ or less. If this is correct replica sends
+  message _VIEW-CHANGE-ACK(j, i, v+1, vd)_ to new primary.
+  
+- New primary collects _VIEW-CHANGE_ + corresponding _n-f-2 VIEW-CHANGE-ACK_ 
+  messages for each replica _i_, which (when considering _VIEW-CHANGE-ACK_ 
+  which new primary could send to itself) form a view-change certificate, so 
+  it can be sure that _n-f_ replicas have same _VIEW-CHANGE_ message for
+  replica _i_. Each view-change certificate is stored in _S_, and when
+  each new view-change cerficate is added new primary makes an attempt to
+  construct _NEW-VIEW_ message. Authors of PBFT state that when new primary
+  gets _n-f_ view-change certificates from normal replicas it is guaranteed
+  that this attempt will succeed.  
+
+- Attempt to form a _NEW-VIEW_ message is as follows:
+  - decide on checkpoint _(ck, cd)_ to use as a starting state for new view:
+    - it should be largest known checkpoint
+    - it should be contained in _C_ of _f+1_ replicas
+    - it should be not less than _h_ of _n-f_ replicas
+  - for each _k_ in _ck+1 .. ck+L_:
+    - for each _(k, d, v)_ in any _P_ from any replica it should get into
+      new view if:
+      - there are _n-f_ replicas with _h < k_ in which for any _(k, d', v')_
+        in _P_ _v' < v_ or _v'=v_, _d'=d_, and  
+      - there are _f+1_ replicas with _Q_ containing _(k, d, v')_, _v' >= v_
+    - if _P_ from _n-f_ replicas with _h < k_ doesn't contain request with
+      this sequence number it was certainly not ordered and authors of PBFT
+      suggest to assign _null_ request to _k_
+
+  Since indy plenum performs dynamic validation of requests before sending
+  or accepting pre-prepare this also should be done for all requests selected
+  for ordering in new view. Those that fail validation were certainly not
+  ordered in previous view so they should be also transformed to _null_
+  requests. Also when doing this validation uncommited state gets updated.
+
+  If we got requests for all _k_ in _ck+1 .. ck+L_ we can send _NEW-VIEW_
+  message. Goal of original approach was to get as many requests to new view 
+  as possible, but if we are okay with discarding requests that were not 
+  commited then we could truncate request list for new view on first _null_ 
+  request, since all later requests were certainly not commited either.
+
+  Side note: authors of PBFT papers don't rely on dynamic validation during
+  pre-prepare as they solve abstract problem of making sure that all replicas
+  of state machine receive same events in same order. If we abstract state
+  from BFT protocol we can just move dynamic validation to execution step and
```

Comment:
What happens if a request fails dynamic validation during execution step? Does request still end up in the ledger? If no then other nodes need to be made aware of which request are being rejected and one more round of consensus (a node with outdated code or other differences can end up with wrong result on the ledger). If yes then we end up with invalid requests in the ledger with some marker indicating invalid requests. The latter is/was done in Fabric and some paper like [this one](https://www.cs.rochester.edu/meetings/sosp2003/papers/p195-yin.pdf).
The objective of doing dynamic validation and request application during consensus that consensus not only guarantees the order of requests but also the final state after the proposal (Pre-Prepare) is processed.

Reply:
@lovesh no, request will not end up in ledger, it's execution won't modify any state, and reply will indicate failure. Nodes with outdated code that process requests differently should be considered malicious, but I see your point that checking this during pre-prepare phase will catch errors earlier.